# Segmenting and Clustering Neighborhoods in Toronto

## 1. Gather Neighborhoods in Toronto

This following Wikipedia page contains postal codes, Borough and Neighborhood information in Toronto, scraping the website use BeautifulSoup to extract the data.

In [202]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

Create soup object of our Wiki page.

In [203]:
html = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source = requests.get(html).text
soup = BeautifulSoup(source, 'lxml')

From the returned html structure, our interested information is in tag 'tr' and 'td', let's extract the contents and make them into a list.

In [204]:
templist = []
for line in soup.find_all('tr'):
    if len(line.find_all('th')) != 0:
        templist.append([item.text.strip('\n') for item in line.find_all('th')])
    else:
        templist.append([item.text.strip('\n') for item in line.find_all('td')])

Convert the templist into a pandas dataframe, and look print out the size

In [205]:
df = pd.DataFrame([item[0:3] for item in templist[1:-5]])
df.columns = templist[0]
print(df.shape)
df.head()

(289, 3)


Postcode           Borough     Neighbourhood
0  M1A      Not assigned      Not assigned    
1  M2A      Not assigned      Not assigned    
2  M3A      North York        Parkwoods       
3  M4A      North York        Victoria Village
4  M5A      Downtown Toronto  Harbourfront

#### Process the dataframe, first remove all records that Borough is 'Not assigned'.

In [206]:
df = df[df.Borough != 'Not assigned'] 
print(df.shape)
df.head()

(212, 3)


Postcode           Borough     Neighbourhood
2  M3A      North York        Parkwoods       
3  M4A      North York        Victoria Village
4  M5A      Downtown Toronto  Harbourfront    
5  M5A      Downtown Toronto  Regent Park     
6  M6A      North York        Lawrence Heights

#### Second, for records that has Borough information, but without Neighbourhood, assign Borough to Neighbourhood.

In [207]:
temp = df[df['Neighbourhood'] =='Not assigned'].Borough
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood']=temp
df = df.reset_index(drop=True)
print(df.shape)
df.head()

(212, 3)


Postcode           Borough     Neighbourhood
0  M3A      North York        Parkwoods       
1  M4A      North York        Victoria Village
2  M5A      Downtown Toronto  Harbourfront    
3  M5A      Downtown Toronto  Regent Park     
4  M6A      North York        Lawrence Heights

#### Third, group records that have the same postcode and Borough into one row with the neighbourhoods, comma delimited.

In [208]:
temp = df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(lambda x: '%s' % ','.join(x))
df1 = pd.DataFrame(temp.index.get_level_values(0))
df2 = pd.DataFrame(temp.index.get_level_values(1))
df3 = pd.DataFrame(temp.values)
df4 = pd.concat([df1, df2, df3], axis=1)
df4.columns = ['Postcode', 'Borough', 'Neighbourhood']
pd.set_option('display.max_colwidth', -1)
# df4[df4.Postcode == 'M9V']


To exam if rows are correctly combined, check one postcode that has different neighbourhoods, M9V.

In [209]:
print(df4[df4.Postcode == 'M9V']['Neighbourhood'])
df[df.Postcode == 'M9V']

101    Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown
Name: Neighbourhood, dtype: object


Postcode    Borough     Neighbourhood
174  M9V      Etobicoke  Albion Gardens  
175  M9V      Etobicoke  Beaumond Heights
176  M9V      Etobicoke  Humbergate      
177  M9V      Etobicoke  Jamestown       
178  M9V      Etobicoke  Mount Olive     
179  M9V      Etobicoke  Silverstone     
180  M9V      Etobicoke  South Steeles   
181  M9V      Etobicoke  Thistletown

Let's see the final data frame size

In [210]:
print(df4.shape)
df4.head()

(103, 3)


Postcode      Borough                         Neighbourhood
0  M1B      Scarborough  Rouge,Malvern                       
1  M1C      Scarborough  Highland Creek,Rouge Hill,Port Union
2  M1E      Scarborough  Guildwood,Morningside,West Hill     
3  M1G      Scarborough  Woburn                              
4  M1H      Scarborough  Cedarbrae

## 2. Assign coordinates to Neighborhoods

Now we have the postal code of each neighborhood along with the borough/neighborhood name, in order to utilize the Foursquare location data, we need to get the Lat, Lon of each neighborhood.

In [211]:
# !conda install -c conda-forge geocoder --yes

In [212]:
import geocoder

def getCoord(postal_code):
    lat_lng_coords = None
    print(postal_code)
    while(lat_lng_coords is None):
        print('looping...')
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
        print(g, g.latlng)
        
    return lat_lng_coord[0], lat_lng_coord[1]
print("Coordinates collector is ready to use!")

Coordinates collector is ready to use!


Because get lat lon from geocoder.google is taking too long for even a single postal code, we are going to use the csv file that already contains coordinates we are looking for.

In [213]:
CoordList = 'Geospatial_Coordinates.csv'
coord_df = pd.read_csv(CoordList)
coord_df.columns = ['Postcode', 'Latitude', 'Longitude']
coord_df.head()

Postcode   Latitude  Longitude
0  M1B      43.806686 -79.194353
1  M1C      43.784535 -79.160497
2  M1E      43.763573 -79.188711
3  M1G      43.770992 -79.216917
4  M1H      43.773136 -79.239476

Now combine this coord_df with previous dataframe we scraped from wikipedia.

In [214]:
toronto_df = pd.merge(df4, coord_df, on='Postcode')
toronto_df.head()

Postcode      Borough                         Neighbourhood   Latitude  \
0  M1B      Scarborough  Rouge,Malvern                         43.806686   
1  M1C      Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2  M1E      Scarborough  Guildwood,Morningside,West Hill       43.763573   
3  M1G      Scarborough  Woburn                                43.770992   
4  M1H      Scarborough  Cedarbrae                             43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

**Create a map of Toronto with neighborhoods superimposed on top.**

In [215]:
import json # library to handle JSON files

# !conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import numpy as np
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [216]:
geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode('Toronto, Ontario')
latitude = location.latitude
longitude = location.longitude
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## 3. Explore Neighborhoods in Toronto

FourSquare API is used here to explore venues and their basic info around each Neighborhood.

First, define FourSquare credentials

In [217]:
CLIENT_ID = 'VTNJI1NBXEBGDETQPO5ITG5ZNRCR0LTLH4BON0L0FMCCQ0YD' # your Foursquare ID
CLIENT_SECRET = 'JCU14KCM31VSSPKWD4H3EFZYB32WD2YBCGCTSM045TANBYII' # your Foursquare Secret
VERSION = '201801122' # Foursquare API version
LIMIT = 500

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VTNJI1NBXEBGDETQPO5ITG5ZNRCR0LTLH4BON0L0FMCCQ0YD
CLIENT_SECRET:JCU14KCM31VSSPKWD4H3EFZYB32WD2YBCGCTSM045TANBYII


#### Let's create a function to repeat the process of exploring venues to all the neighborhoods in Toronto, radius is set to 1500 meters to ensure each neighborhood could get at least one venue nearby

In [218]:
def getNearbyVenues(names, latitudes, longitudes, radius=1500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [220]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Take a look at returned venues dataframe of Toronto.

In [221]:
print(toronto_venues.shape)
toronto_venues.head()

(6747, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Rouge,Malvern  43.806686             -79.194353                
1  Rouge,Malvern  43.806686             -79.194353                
2  Rouge,Malvern  43.806686             -79.194353                
3  Rouge,Malvern  43.806686             -79.194353                
4  Rouge,Malvern  43.806686             -79.194353                

                Venue  Venue Latitude  Venue Longitude        Venue Category  
0  Images Salon & Spa  43.802283      -79.198565        Spa                   
1  Canadiana exhibit   43.817962      -79.193374        Zoo Exhibit           
2  Caribbean Wave      43.798558      -79.195777        Caribbean Restaurant  
3  Harvey's            43.800106      -79.198258        Fast Food Restaurant  
4  Wendy's             43.802008      -79.198080        Fast Food Restaurant

6747 venues in total were found, let's see how many venues were returned for each neighborhood, and how many unique categories among them.

In [222]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                                                                    
Adelaide,King,Richmond                                                                                  100                     
Agincourt                                                                                               59                      
Agincourt North,L'Amoreaux East,Milliken,Steeles East                                                   73                      
Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown  29                      
Alderwood,Long Branch                                                                                   47                      
Bathurst Manor,Downsview North,Wilson Heights                                                           42                      
Bayview Village                                                                                         15                      
Bedford Park,Lawrence Manor East                                                                        75                      
Berczy Park                                                                                             100                     
Birch Cliff,Cliffside West                                                                              14                      
Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe                                              38                      
Brockton,Exhibition Place,Parkdale Village                                                              100                     
Business reply mail Processing Centre969 Eastern                                                        100                     
CFB Toronto,Downsview East                                                                              30                      
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara    61                      
Cabbagetown,St. James Town                                                                              100                     
Caledonia-Fairbanks                                                                                     67                      
Canada Post Gateway Processing Centre                                                                   54                      
Cedarbrae                                                                                               64                      
Central Bay Street                                                                                      100                     
Chinatown,Grange Park,Kensington Market                                                                 100                     
Christie                                                                                                100                     
Church and Wellesley                                                                                    100                     
Clairlea,Golden Mile,Oakridge                                                                           39                      
Clarks Corners,Sullivan,Tam O'Shanter                                                                   46                      
Cliffcrest,Cliffside,Scarborough Village West                                                           41                      
Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park                                      20                      
Commerce Court,Victoria Hotel                                                                           100                     
Davisville                                                                                              100                     
Davisville North                                                                                        1

In [223]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 335 unique categories.


Check if we successfully explored all neighborhoods, there should be 103 neighborhoods.

In [224]:
print('There are {} unique neighborhood.'.format(len(toronto_venues['Neighborhood'].unique())))

There are 103 unique neighborhood.


## 4. Analyze Each Neighborhood

#### Encoding categorized value 'Venue Catgory'

In [225]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Rouge,Malvern  0                  0               0                   
1  Rouge,Malvern  0                  0               0                   
2  Rouge,Malvern  0                  0               0                   
3  Rouge,Malvern  0                  0               0                   
4  Rouge,Malvern  0                  0               0                   

   African Restaurant  Airport  Airport Lounge  American Restaurant  \
0  0                   0        0               0                     
1  0                   0        0               0                     
2  0                   0        0               0                     
3  0                   0        0               0                     
4  0                   0        0               0                     

   Amphitheater  Animal Shelter     ...       Warehouse Store  Whisky Bar  \
0  0             0                  ...       0                0            
1  0             0                  ...       0                0            
2  0             0                  ...       0                0            
3  0             0                  ...       0                0            
4  0             0                  ...       0                0            

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Xinjiang Restaurant  \
0  0         0          0            0              0                     
1  0         0          0            0              0                     
2  0         0          0            0              0                     
3  0         0          0            0              0                     
4  0         0          0            0              0                     

   Yoga Studio  Zoo  Zoo Exhibit  
0  0            0    0            
1  0            0    1            
2  0            0    0            
3  0            0    0            
4  0            0    0            

[5 rows x 336 columns]

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [226]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
print(toronto_grouped.shape)
toronto_grouped.head()

(103, 336)


Neighbourhood  \
0  Adelaide,King,Richmond                                                                                   
1  Agincourt                                                                                                
2  Agincourt North,L'Amoreaux East,Milliken,Steeles East                                                    
3  Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown   
4  Alderwood,Long Branch                                                                                    

   Accessories Store  Adult Boutique  Afghan Restaurant  African Restaurant  \
0  0.0                0.0             0.0                0.0                  
1  0.0                0.0             0.0                0.0                  
2  0.0                0.0             0.0                0.0                  
3  0.0                0.0             0.0                0.0                  
4  0.0                0.0             0.0                0.0                  

   Airport  Airport Lounge  American Restaurant  Amphitheater  Animal Shelter  \
0  0.0      0.0             0.02                 0.0           0.0              
1  0.0      0.0             0.00                 0.0           0.0              
2  0.0      0.0             0.00                 0.0           0.0              
3  0.0      0.0             0.00                 0.0           0.0              
4  0.0      0.0             0.00                 0.0           0.0              

      ...       Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0     ...       0.0              0.0         0.0       0.0        0.000000      
1     ...       0.0              0.0         0.0       0.0        0.000000      
2     ...       0.0              0.0         0.0       0.0        0.000000      
3     ...       0.0              0.0         0.0       0.0        0.000000      
4     ...       0.0              0.0         0.0       0.0        0.021277      

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo  Zoo Exhibit  
0  0.0            0.000000             0.0          0.0  0.0          
1  0.0            0.016949             0.0          0.0  0.0          
2  0.0            0.000000             0.0          0.0  0.0          
3  0.0            0.000000             0.0          0.0  0.0          
4  0.0            0.000000             0.0          0.0  0.0          

[5 rows x 336 columns]

#### Let's get each neighborhood along with the top 10 most common venues

In [227]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]


In [228]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)
    
    neighborhoods_venues_sorted

neighborhoods_venues_sorted

Neighborhood  \
0    Adelaide,King,Richmond                                                                                   
1    Agincourt                                                                                                
2    Agincourt North,L'Amoreaux East,Milliken,Steeles East                                                    
3    Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown   
4    Alderwood,Long Branch                                                                                    
5    Bathurst Manor,Downsview North,Wilson Heights                                                            
6    Bayview Village                                                                                          
7    Bedford Park,Lawrence Manor East                                                                         
8    Berczy Park                                                                                              
9    Birch Cliff,Cliffside West                                                                               
10   Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe                                               
11   Brockton,Exhibition Place,Parkdale Village                                                               
12   Business reply mail Processing Centre969 Eastern                                                         
13   CFB Toronto,Downsview East                                                                               
14   CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara     
15   Cabbagetown,St. James Town                                                                               
16   Caledonia-Fairbanks                                                                                      
17   Canada Post Gateway Processing Centre                                                                    
18   Cedarbrae                                                                                                
19   Central Bay Street                                                                                       
20   Chinatown,Grange Park,Kensington Market                                                                  
21   Christie                                                                                                 
22   Church and Wellesley                                                                                     
23   Clairlea,Golden Mile,Oakridge                                                                            
24   Clarks Corners,Sullivan,Tam O'Shanter                                                                    
25   Cliffcrest,Cliffside,Scarborough Village West                                                            
26   Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park                                       
27   Commerce Court,Victoria Hotel                                                                            
28   Davisville                                                                                               
29   Davisville North                                                                                         
..                ...                                                                                         
73   Parkdale,Roncesvalles                                                                                    
74   Parkwoods                                                                                                
75   Queen's Park                                                                                             
76   Rosedale                                                                                                 
77   Roselawn                                                                                      

## 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters

In [229]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

len(kmeans.labels_)

103

#### Let's visualize the resulting clusters.

In [230]:
toronto_merged = toronto_df
toronto_merged['Neighborhood'] = toronto_merged['Neighbourhood']
toronto_merged['Cluster Labels'] = kmeans.labels_
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_grouped_clustering.shape


(103, 335)

In [231]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start = 10)

x = np.arange(kclusters)
ys = [i+x+(i+x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon], 
        radius = 5,
        popup = label, 
        color = rainbow[cluster-1],
        fill = True,
        fill_color = rainbow[cluster-1],
        fill_opacity = 0.7).add_to(map_clusters)

map_clusters

## 6. Examine Clusters

Now, let's examine each cluster and determine the discriminating venue categories that distinguish each cluster.

In [232]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  \
0   Scarborough        
8   Scarborough        
11  Scarborough        
12  Scarborough        
15  Scarborough        
19  North York         
20  North York         
21  North York         
22  North York         
27  North York         
28  North York         
30  North York         
32  North York         
35  East York          
40  East York          
43  East Toronto       
48  Central Toronto    
49  Central Toronto    
50  Downtown Toronto   
51  Downtown Toronto   
65  Central Toronto    
73  York               
75  Downtown Toronto   
76  West Toronto       
79  North York         
80  York               
83  West Toronto       
84  West Toronto       
85  Queen's Park       
86  Mississauga        
88  Etobicoke          
89  Etobicoke          

                                                     Neighborhood  \
0   Rouge,Malvern                                                   
8   Cliffcrest,Cliffside,Scarborough Village West                   
11  Maryvale,Wexford                                                
12  Agincourt                                                       
15  L'Amoreaux West,Steeles West                                    
19  Bayview Village                                                 
20  Silver Hills,York Mills                                         
21  Newtonbrook,Willowdale                                          
22  Willowdale South                                                
27  Flemingdon Park,Don Mills South                                 
28  Bathurst Manor,Downsview North,Wilson Heights                   
30  CFB Toronto,Downsview East                                      
32  Downsview Central                                               
35  Woodbine Gardens,Parkview Hill                                  
40  East Toronto                                                    
43  Studio District                                                 
48  Moore Park,Summerhill East                                      
49  Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West   
50  Rosedale                                                        
51  Cabbagetown,St. James Town                                      
65  The Annex,North Midtown,Yorkville                               
73  Humewood-Cedarvale                                              
75  Christie                                                        
76  Dovercourt Village,Dufferin                                     
79  Maple Leaf Park,North Park,Upwood Park                          
80  Del Ray,Keelesdale,Mount Dennis,Silverthorn                     
83  Parkdale,Roncesvalles                                           
84  Runnymede,Swansea                                               
85  Queen's Park                                                    
86  Canada Post Gateway Processing Centre                           
88  Humber Bay Shores,Mimico South,New Toronto                      
89  Alderwood,Long Branch                                           

    Cluster Labels      1st Most Common Venue   2nd Most Common Venue  \
0   0               Zoo Exhibit                Fast Food Restaurant     
8   0               Park                       Harbor / Marina          
11  0               Middle Eastern Restaurant  Coffee Shop              
12  0               Chinese Restaurant         Coffee Shop              
15  0               Chinese Restaurant         Coffee Shop              
19  0               Japanese Restaurant        Bank                     
20  0               Coffee Shop                Furniture / Home Store   
21  0               Korean Restaurant          Coffee Shop              
22  0               Korean Restaurant          Coffee Shop              
27  0               Coffee Shop                Japanese Restaurant      
28  0               Park                       Coffee Shop              
30  0               Athletics & Sports         Racetrack          

#### Number of international restaurants is outstanding in category 1 neighborhood, therefore this category 1 could named as International Neighborhood. People who love the taste of exotic or crave of a bite from their home countries are most attracted here.

In [233]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  \
4    Scarborough        
7    Scarborough        
16   Scarborough        
17   North York         
18   North York         
23   North York         
29   North York         
31   North York         
33   North York         
41   East Toronto       
42   East Toronto       
44   Central Toronto    
45   Central Toronto    
52   Downtown Toronto   
55   Downtown Toronto   
57   Downtown Toronto   
60   Downtown Toronto   
62   North York         
63   Central Toronto    
64   Central Toronto    
67   Downtown Toronto   
68   Downtown Toronto   
69   Downtown Toronto   
70   Downtown Toronto   
71   North York         
72   North York         
77   West Toronto       
78   West Toronto       
82   West Toronto       
87   East Toronto       
90   Etobicoke          
91   Etobicoke          
92   Etobicoke          
96   North York         
97   North York         
101  Etobicoke          
102  Etobicoke          

                                                                                                                        Neighborhood  \
4    Cedarbrae                                                                                                                         
7    Clairlea,Golden Mile,Oakridge                                                                                                     
16   Upper Rouge                                                                                                                       
17   Hillcrest Village                                                                                                                 
18   Fairview,Henry Farm,Oriole                                                                                                        
23   York Mills West                                                                                                                   
29   Northwood Park,York University                                                                                                    
31   Downsview West                                                                                                                    
33   Downsview Northwest                                                                                                               
41   The Danforth West,Riverdale                                                                                                       
42   The Beaches West,India Bazaar                                                                                                     
44   Lawrence Park                                                                                                                     
45   Davisville North                                                                                                                  
52   Church and Wellesley                                                                                                              
55   St. James Town                                                                                                                    
57   Central Bay Street                                                                                                                
60   Design Exchange,Toronto Dominion Centre                                                                                           
62   Bedford Park,Lawrence Manor East                                                                                                  
63   Roselawn                                                                                                                          
64   Forest Hill North,Forest Hill West                                                                                                
67   Chinatown,Grange Park,Kensington Market                                                                                           
68   CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and S

#### Category 2 has plenty of coffee shop, spa and Bar, we named it Leisure Neighborhood, these are nice area to enjoy a nice layback morning or hangout with a couple friends.

In [234]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough      Neighborhood  Cluster Labels 1st Most Common Venue  \
93  Etobicoke  Islington Avenue  2               Grocery Store          

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common  \
93  Pharmacy              Convenience Store     Japanese Restaurant   

   5th Most Common 6th Most Common 7th Most Common 8th Most Common  \
93  Bus Line        Café            Shopping Mall   Liquor Store     

   9th Most Common 10th Most Common  
93  Golf Course     Bank

#### Category 3's top three venues are grocery store, parmacy and convenience store, named as 'Shopping Neighborhood', very convenient neighborhoods that everything you need is just a couple minutes away.

In [235]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  \
1    Scarborough        
2    Scarborough        
3    Scarborough        
24   North York         
34   North York         
36   East York          
37   East Toronto       
39   East York          
46   Central Toronto    
47   Central Toronto    
53   Downtown Toronto   
59   Downtown Toronto   
61   Downtown Toronto   
66   Downtown Toronto   
74   York               
81   York               
94   Etobicoke          
95   Etobicoke          
98   York               
99   Etobicoke          
100  Etobicoke          

                                                             Neighborhood  \
1    Highland Creek,Rouge Hill,Port Union                                   
2    Guildwood,Morningside,West Hill                                        
3    Woburn                                                                 
24   Willowdale West                                                        
34   Victoria Village                                                       
36   Woodbine Heights                                                       
37   The Beaches                                                            
39   Thorncliffe Park                                                       
46   North Toronto West                                                     
47   Davisville                                                             
53   Harbourfront,Regent Park                                               
59   Harbourfront East,Toronto Islands,Union Station                        
61   Commerce Court,Victoria Hotel                                          
66   Harbord,University of Toronto                                          
74   Caledonia-Fairbanks                                                    
81   The Junction North,Runnymede                                           
94   Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park     
95   Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe             
98   Weston                                                                 
99   Westmount                                                              
100  Kingsview Village,Martin Grove Gardens,Richview Gardens,St. Phillips   

     Cluster Labels 1st Most Common Venue          2nd Most Common Venue  \
1    3               Gym                   Breakfast Spot                  
2    3               Pizza Place           Coffee Shop                     
3    3               Coffee Shop           Pharmacy                        
24   3               Coffee Shop           Pizza Place                     
34   3               Coffee Shop           Middle Eastern Restaurant       
36   3               Pizza Place           Coffee Shop                     
37   3               Coffee Shop           Pub                             
39   3               Coffee Shop           Sandwich Place                  
46   3               Coffee Shop           Italian Restaurant              
47   3               Coffee Shop           Italian Restaurant              
53   3               Coffee Shop           Café                            
59   3               Hotel                 Coffee Shop                     
61   3               Coffee Shop           Hotel                           
66   3               Café                  Vegetarian / Vegan Restaurant   
74   3               Coffee Shop           Italian Restaurant              
81   3               Coffee Shop           Grocery Store                   
94   3               Park                  Convenience Store               
95   3               Baseball Field        Convenience Store               
98   3               Grocery Store         Bakery                          
99   3               Pizza Place           Coffee Shop                     
100  3               Pizza Place           Coffee Shop                     

    3rd Most Common Venue         4th Most Common            5th Most Common  \
1    Neighborho

#### Category 4 is a "Balanced Neighborhood", in these neighborhood you likely to find varies types of venues that fit different services, such as Park, Gym, Bank, etc.

In [236]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough                                           Neighborhood  \
5   Scarborough       Scarborough Village                                     
6   Scarborough       East Birchmount Park,Ionview,Kennedy Park               
9   Scarborough       Birch Cliff,Cliffside West                              
10  Scarborough       Dorset Park,Scarborough Town Centre,Wexford Heights     
13  Scarborough       Clarks Corners,Sullivan,Tam O'Shanter                   
14  Scarborough       Agincourt North,L'Amoreaux East,Milliken,Steeles East   
25  North York        Parkwoods                                               
26  North York        Don Mills North                                         
38  East York         Leaside                                                 
54  Downtown Toronto  Ryerson,Garden District                                 
56  Downtown Toronto  Berczy Park                                             
58  Downtown Toronto  Adelaide,King,Richmond                                  

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5   4               Sandwich Place        Grocery Store          
6   4               Coffee Shop           Fast Food Restaurant   
9   4               Park                  Restaurant             
10  4               Coffee Shop           Fast Food Restaurant   
13  4               Chinese Restaurant    Fast Food Restaurant   
14  4               Chinese Restaurant    BBQ Joint              
25  4               Coffee Shop           Park                   
26  4               Japanese Restaurant   Coffee Shop            
38  4               Coffee Shop           Indian Restaurant      
54  4               Coffee Shop           Gastropub              
56  4               Café                  Coffee Shop            
58  4               Coffee Shop           Hotel                  

   3rd Most Common Venue     4th Most Common        5th Most Common  \
5   Pharmacy              Coffee Shop         Fast Food Restaurant    
6   Sandwich Place        Chinese Restaurant  Pizza Place             
9   Golf Course           Gym                 General Entertainment   
10  Pizza Place           Grocery Store       Chinese Restaurant      
13  Park                  Pharmacy            Falafel Restaurant      
14  Bakery                Coffee Shop         Korean Restaurant       
25  Bank                  Supermarket         Bus Stop                
26  Burger Joint          Park                Bank                    
38  Restaurant            Supermarket         Bakery                  
54  American Restaurant   Diner               Cosmetics Shop          
56  Hotel                 Restaurant          Park                    
58  Café                  Gastropub           Theater                 

        6th Most Common                7th Most Common  \
5   Ice Cream Shop       Wings Joint                     
6   Grocery Store        Train Station                   
9   Skating Rink         Diner                           
10  Indian Restaurant    Wings Joint                     
13  Pizza Place          Deli / Bodega                   
14  Pizza Place          Noodle House                    
25  Pharmacy             Paper / Office Supplies Store   
26  Pizza Place          Restaurant                      
38  Burger Joint         Grocery Store                   
54  Tea Room             Restaurant                      
56  Japanese Restaurant  Gastropub                       
58  Pizza Place          Thai Restaurant                 

              8th Most Common       9th Most Common       10th Most Common  
5   Chinese Restaurant         Pizza Place           Liquor Store           
6   Pharmacy                   Sporting Goods Shop   Beer Store             
9   Café                       Thai Restaurant       Bank                   
10  Pet Store                  Light Rail Station    Burger Joint           
13  Sandwich Place             Cantonese Restaurant  Vietnames

#### Category 5 is 'Busy-commuter Neighborhood', as the top number venues are fast food restaurants, pizza place, Burger Joint, etc.